Add medal emojis to top 3 scores

jgdhs27 opened the pull request:
Add medal emojis (🥇, 🥈, 🥉) to the top 3 scores for a certain difficulty and shot.

![image](https://user-images.githubusercontent.com/44336657/233852541-f2f56dc3-2d91-4ab3-9a3d-634b3ec9f2ab.png)

In the future, I would like to add a player ranking by medal count, like royalflare had.

n-rook commented on project/thscoreboard/replays/replays_to_json.py:
```diff
@@ -63,8 +77,21 @@ def _get_th17_additional_fields(self, shot: models.Shot) -> dict:
+
     def convert_replays_to_serializable_list(
         self,
         replays: Iterable[models.Replay],
```
seems like we've introduced an undocumented requirement that replays be ordered (in reverse order) by score, no?

jgdhs27 commented:
Ok, I came up with another approach which adds the ranking inside the db query. I find this cleaner, I hope it addresses your concerns.

jgdhs27 commented:
Also, this makes it so that I could show the medals on the user page and front page without much difficulty, so I've done that.

n-rook commented on project/thscoreboard/replays/replays_to_json.py:
```diff
@@ -65,6 +69,28 @@ def _get_th17_additional_fields(self, shot: models.Shot) -> dict:
+
+def add_rank_annotation_to_replays(
+    replays: Manager[models.Replay],
+) -> Manager[models.Replay]:
```
Could you move this function over to models.py? We may want in places that aren't returning json, but it'd be strange to import "replays_to_json" in a place where we aren't converting replays to json. Plus, looking at the signature, I think you can define this as a method on a custom manager, if I remember Django right. Certainly calling replays.objects.some_query().annotate_with_rank() would be very convenient.

jgdhs27 replied:
Done. Although I think the current implementation isnt perfect - I can only call `.annotate_with_rank()` on a `ReplayQuerySet` object but not a `ReplayManager` object, which feels wrong.

jgdhs27 replied:
disregard that, fixed that little issue too.

n-rook reviewed (changes requested):
Nice, this definitely looks like a good way to do it. Just one more comment

jgdhs27 commented:
Ok, another revision done. `annotate_with_rank` is now part of the `ReplayQuerySet` object, leading to much nicer code. Added tests to that. I also fixed a bug where TASes and irregular replays were counted in the rankings.

n-rook reviewed (approved):
Nice!

I'm _slightly_ worried about performance, but it's probably fine :)
